devices/oneplus-enchilada/default.nix

```diff
@@ -0,0 +1,126 @@
+    firmware = [
+      config.mobile.device.firmware
+    ];
+  };
+
+
+  mobile.device.firmware = pkgs.callPackage ./firmware {};
+
+  mobile.system.android.device_name = "OnePlus6";
+  mobile.system.android = {
+    ab_partitions = true;
+    bootimg.flash = {
+      offset_base = "0x00000000";
+      offset_kernel = "0x00008000";
+      offset_ramdisk = "0x01000000";
+      offset_second = "0x00f00000";
+      offset_tags = "0x00000100";
+      pagesize = "4096";
+    };
+  };
+
+  boot.kernelParams = [
+    "console=tty0"
+    "panic=10"
+  ];
+
+  mobile.usb.mode = "gadgetfs";
+  # Google
+  mobile.usb.idVendor = "18D1";
+  # "Nexus 4"
+  mobile.usb.idProduct = "D001";
+
+  mobile.system.type = "android";
+
+  hardware.enableRedistributableFirmware = true;
+  hardware.firmware = lib.mkBefore [ config.mobile.device.firmware ];
+
+
+  mobile.usb.gadgetfs.functions = {
+    adb = "ffs.adb";
+    rndis = "rndis.usb0";
+  };
+
+  systemd.services = {
+    rmtfs = rec {
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "qrtr-ns.service" ];
+      after = requires;
+      serviceConfig = {
+        ExecStart = "${pkgs.rmtfs}/bin/rmtfs -r -P -s";
+        Restart = "always";
+        RestartSec = "1";
+      };
+    };
+    qrtr-ns = rec {
+      serviceConfig = {
+        ExecStart = "${pkgs.qrtr}/bin/qrtr-ns -f 1";
+        Restart = "always";
+      };
+    };
+    tqftpserv = rec {
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "qrtr-ns.service" ];
+      after = requires;
+      serviceConfig = {
+        ExecStart = "${pkgs.tqftpserv}/bin/tqftpserv";
+        Restart = "always";
+      };
+    };
+    pd-mapper = rec {
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "qrtr-ns.service" ];
+      after = requires;
+      serviceConfig = {
+        ExecStart = "${pkgs.pd-mapper}/bin/pd-mapper";
+        Restart = "always";
+      };
+    };
+    msm-modem-uim-selection = {
+      enable = true;
+      before = [ "ModemManager.service" ];
+      wantedBy = [ "ModemManager.service" ];
+      path = with pkgs; [ libqmi gawk gnugrep ];
+      script = ''
+        QMICLI_MODEM="qmicli --silent -pd qrtr://0"
+        QMI_CARDS=$($QMICLI_MODEM --uim-get-card-status)
+        if ! printf "%s" "$QMI_CARDS" | grep -Fq "Primary GW:   session doesn't exist"
+        then
+            $QMICLI_MODEM --uim-change-provisioning-session='activate=no,session-type=primary-gw-provisioning' > /dev/null
+        fi
+        FIRST_PRESENT_SLOT=$(printf "%s" "$QMI_CARDS" | grep "Card state: 'present'" -m1 -B1 | head -n1 | cut -c7-7)
+        FIRST_PRESENT_AID=$(printf "%s" "$QMI_CARDS" | grep "usim (2)" -m1 -A3 | tail -n1 | awk '{print $1}')
+        $QMICLI_MODEM --uim-change-provisioning-session="slot=$FIRST_PRESENT_SLOT,activate=yes,session-type=primary-gw-provisioning,aid=$FIRST_PRESENT_AID" > /dev/null
+      '';
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+      };
+    };
+  };
```

Comment:
Prefer defining those that are not device-specific as "quirks".

(The "quirk" naming may change in the future, especially to distinguish between "quirky behaviour" and "normal config" use.)

Here's a quirk example:

 - https://github.com/NixOS/mobile-nixos/blob/8a105e177632f0fbc4ca28ee0195993baf0dcf9a/modules/quirks/qualcomm/wcnss-wlan.nix

It defines an enable option, which can be opted-into by the device config.

I don't know if there's a modem "generation name" to use, so instead `quirks.qualcomm.mainline.sdm845-modem.enable` would work, and modems that use the same exact config could add an option for the SoC enabling the same options.

If you need more help/clarifications, don't hesitate to ask!